Return errors from importmulti if complete rescans are not successful

ryanofsky opened the pull request:
This is supposed to address the concern from @gmaxwell in #9761 that importmulti rescans on pruned nodes could unsuccessfully scan pruned blocks and fail to inform the user.

An alternative approach of just simply disabling importmulti rescans on pruned nodes was initially discussed in https://botbot.me/freenode/bitcoin-core-dev/msg/81023921/, but running being able to import keys and do partial rescans on pruned nodes was determined to be an important use-case.

~~~Posting this PR for feedback. The code has not been tested at all, and I'm actually not sure how I'm going to go about writing a unit test because there doesn't seem to be an easy way to target individual blocks for pruning inside RPC tests.~~~

morcos commented:
Concept ACK
This looks like a reasonable approach to me.  

Don't forget that you may want to include the grace period in detecting whether it scanned enough.
Also you might want to skip the rest of the loop if ReadBlockFromDisk is false.

laanwj commented:
Concept ACK.

> The code has not been tested at all, and I'm actually not sure how I'm going to go about writing a unit test because there doesn't seem to be an easy way to target individual blocks for pruning inside RPC tests.

There's `pruneblockchain` but yes pruning works on a per-file, not a per-block basis, so it would need some extra effort on regtest to generate full blocks instead of empty ones otherwise at block 30000 you're still in the first file, and that RPC will do nothing.

gmaxwell commented:
Concept ACK. Will test/review further.

morcos commented:
Please rebase on #9761 so we can verify the grace period works properly

ryanofsky committed:
Return errors from importmulti if complete rescans are not successful

ryanofsky commented:
Rebased

morcos commented on src/wallet/wallet.cpp:
```diff
@@ -1570,12 +1574,15 @@ int CWallet::ScanForWalletTransactions(CBlockIndex* pindexStart, bool fUpdate)
             {
-                if (AddToWalletIfInvolvingMe(*block.vtx[posInBlock], pindex, posInBlock, fUpdate))
-                    ret++;
+                AddToWalletIfInvolvingMe(*block.vtx[posInBlock], pindex, posInBlock, fUpdate);
```
I think this loop should also be skipped if ReadBlockFromDisk fails

ryanofsky replied:
Skipped in fixmed6f6a23c7092aa21110cc77ebeb018492bacae28

ryanofsky reviewed (commented):
Add test for CWallet::ScanForWalletTransactions in e3d4f257ce0202811464d3b6860db87ba147aa24

jonasschnelli commented on src/wallet/rpcdump.cpp:
```diff
@@ -1072,12 +1073,32 @@ UniValue importmulti(const JSONRPCRequest& mainRequest)
+                } else {
+                    UniValue result = UniValue(UniValue::VOBJ);
+                    result.pushKV("success", UniValue(false));
+                    result.pushKV("error", JSONRPCError(RPC_MISC_ERROR, strprintf("Failed to rescan before time %" PRId64 ", transactions may be missing.", (long long)scannedRange->GetBlockTimeMax())));
```
just curios: Is there an advantage using `PRId64` over `%ul` (rest of the code uses this)?

ryanofsky replied:
Actually I need to change this to either remove PRId64 or the (long long) cast.

You're supposed to use PRId64 whenever you pass an int64_t type to printf. %ld or %lld aren't always correct because longs and long longs just have minimum widths in standard c++, not fixed widths that are the same on all systems.

ryanofsky replied:
Removed cast in e49f0394f07a95e6cc5cdabe32f2c6b8f53c40fb.

ryanofsky commented:
Previously squashed e3d4f257ce0202811464d3b6860db87ba147aa24 -> 455b8f7cb2a1889053a006055e9dcad923c77029 ([multicheck.2](https://github.com/ryanofsky/bitcoin/commits/pr/multicheck.2) -> [multicheck.3](https://github.com/ryanofsky/bitcoin/commits/pr/multicheck.3))

ryanofsky commented:
Added C++ test for new importmulti RPC code in e64cdfba4a7e8a670e76c7bf770c248dc6b78c3d

Squashed e64cdfba4a7e8a670e76c7bf770c248dc6b78c3d -> b73252de549984a06aae294252d0d055aca43a4a ([multicheck.4](https://github.com/ryanofsky/bitcoin/commits/pr/multicheck.4) -> [multicheck.5](https://github.com/ryanofsky/bitcoin/commits/pr/multicheck.5))

sipa commented:
Note that our strprintf is a wrapper around tfm::format, and doesn't need type specs at all.

morcos commented:
lightly tested ACK b73252d
Only reviewed/tested code, not the tests.

jonasschnelli reviewed (approved):
utACK b73252de549984a06aae294252d0d055aca43a4a

laanwj commented on src/wallet/rpcdump.cpp:
```diff
@@ -1072,12 +1073,33 @@ UniValue importmulti(const JSONRPCRequest& mainRequest)
+                } else {
+                    UniValue result = UniValue(UniValue::VOBJ);
+                    result.pushKV("success", UniValue(false));
+                    result.pushKV("error", JSONRPCError(RPC_MISC_ERROR, strprintf("Failed to rescan before time %" PRId64 ", transactions may be missing.", scannedRange->GetBlockTimeMax())));
```
Please remove the PRI64d here. It gives problems with some platforms, and as @sipa ways don't need it.

ryanofsky replied:
Ok will change, but what are the problems with some platforms?

ryanofsky replied:
Squashed b73252de549984a06aae294252d0d055aca43a4a -> 90671ee7e77c17c18f5ea82ea041962ef5d53f8a ([multicheck.5](https://github.com/ryanofsky/bitcoin/commits/pr/multicheck.5) -> [multicheck.6](https://github.com/ryanofsky/bitcoin/commits/pr/multicheck.6))

ryanofsky replied:
Removed in aaaf65a90b87c18889d177ff3f0ab8a868f849f3.

Reason I am asking about the problem you're seeing is that not using PRI64d for int64_t would prevent us from using compile time format string checking (as suggested in https://github.com/bitcoin/bitcoin/issues/9423#issuecomment-269480805) which would be much better than the runtime checking we currently use.

laanwj replied:
> Ok will change, but what are the problems with some platforms?

The conceptual problem is that PRI64d can evaluate to anything depending on the platform, e.g. "%ponyfood". `tinyformat.h` cannot know about all these platform-dependent escape sequences, so might produce garbled output or even raise an exception. We had a lot of problems with this back in the day with windows and mingw.

Offline checking could still work, the script would just need to be smart enough to cope with a) translations b) tinyformat's specifiers and type-safety. (b) is actually a subset, not a superset of what platform printfs support (none of the insecure ones like %n) so if anything that should make it easier.

ryanofsky commented:
Rebased b73252de549984a06aae294252d0d055aca43a4a -> 90671ee7e77c17c18f5ea82ea041962ef5d53f8a ([multicheck.5](https://github.com/ryanofsky/bitcoin/commits/pr/multicheck.5) -> [multicheck.6](https://github.com/ryanofsky/bitcoin/commits/pr/multicheck.6))

ryanofsky commented:
Squashed aaaf65a90b87c18889d177ff3f0ab8a868f849f3 -> 1425c727863b335bf98bb4b17b575e59552d38e9 ([multicheck.7](https://github.com/ryanofsky/bitcoin/commits/pr/multicheck.7) -> [multicheck.8](https://github.com/ryanofsky/bitcoin/commits/pr/multicheck.8))

MarcoFalke commented:
Somehow travis didn't like your commit.

On Fri, Feb 17, 2017 at 1:25 PM, Russell Yanofsky <notifications@github.com>
wrote:

> Squashed aaaf65a
> <https://github.com/bitcoin/bitcoin/commit/aaaf65a90b87c18889d177ff3f0ab8a868f849f3>
> -> 1425c72
> <https://github.com/bitcoin/bitcoin/commit/1425c727863b335bf98bb4b17b575e59552d38e9>
> (multicheck.7
> <https://github.com/ryanofsky/bitcoin/commits/pr/multicheck.7> ->
> multicheck.8
> <https://github.com/ryanofsky/bitcoin/commits/pr/multicheck.8>)
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/bitcoin/bitcoin/pull/9773#issuecomment-280636020>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AGGmvze2Mxoj5bq4mVJGQPY4YpClnvOlks5rdZGogaJpZM4MCUp0>
> .
>

ryanofsky commented:
> Somehow travis didn't like your commit.

Appears to be a spurious error. The travis run for b72521dd7f570b0c557439d534f7523a5be52999 which is aaaf65a90b87c18889d177ff3f0ab8a868f849f3 merged into 9828f9a9962c1bee5c343847030b9cfd87a40a5e succeeded: https://travis-ci.org/bitcoin/bitcoin/builds/202600321

Only the travis run for 64072b74ebbf7b5585b39f6c117c6ffa9dc820f0 which is 1425c727863b335bf98bb4b17b575e59552d38e9 merged into 9828f9a9962c1bee5c343847030b9cfd87a40a5e failed: https://travis-ci.org/bitcoin/bitcoin/builds/202600549

There are no differences between the two commits:

```
$ git rev-parse aaaf65a90b87c18889d177ff3f0ab8a868f849f3^{tree} 1425c727863b335bf98bb4b17b575e59552d38e9^{tree}
9e695fd32f65b1596f3ff613b96ec756bb2f1d2b
9e695fd32f65b1596f3ff613b96ec756bb2f1d2b
```

ryanofsky commented:
Touched CommitDate to rerun travis 1425c727863b335bf98bb4b17b575e59552d38e9 -> 68da3ee4284a07f695e992afd6cdd545604aac90 ([multicheck.8](https://github.com/ryanofsky/bitcoin/commits/pr/multicheck.8) -> [multicheck.9](https://github.com/ryanofsky/bitcoin/commits/pr/multicheck.9))

ryanofsky commented:
Travis run for 91b1ec2f4f27a817aee15f573afc26b7d17c8d7b which is 154025927dd5929f69f79e32818932d387c004bf merged into 9828f9a9962c1bee5c343847030b9cfd87a40a5e succeeded: https://travis-ci.org/bitcoin/bitcoin/builds/202655068

Travis run for 527205931a636f26d620af25db61dbe9c716fb38 which is 68da3ee4284a07f695e992afd6cdd545604aac90 merged into 9828f9a9962c1bee5c343847030b9cfd87a40a5e failed: https://travis-ci.org/bitcoin/bitcoin/builds/202665925

Touched CommitDate to rerun travis 68da3ee4284a07f695e992afd6cdd545604aac90 -> 99d032e5b6a5140cdb07d0d209ff7f6a74bc1244 ([multicheck.9](https://github.com/ryanofsky/bitcoin/commits/pr/multicheck.9) -> [multicheck.10](https://github.com/ryanofsky/bitcoin/commits/pr/multicheck.10))

jtimon commented:
Concept ACK

paveljanik commented on src/wallet/test/wallet_tests.cpp:
```diff
@@ -355,4 +361,58 @@ BOOST_AUTO_TEST_CASE(ApproximateBestSubset)
+    // Verify ScanForWalletTransactions picks up transactions in both the old
+    // and new block files.
+    {
+        CWallet wallet;
```
This brings a few Wshadow warinngs:

```
Wshadow statistics: 
   1 wallet/test/wallet_tests.cpp:377:17: warning: declaration shadows a variable in namespace 'wallet_tests' [-Wshadow]
   1 wallet/test/wallet_tests.cpp:391:17: warning: declaration shadows a variable in namespace 'wallet_tests' [-Wshadow]
   1 wallet/test/wallet_tests.cpp:399:17: warning: declaration shadows a variable in namespace 'wallet_tests' [-Wshadow]

```
